### ISSUE TITLE
DeveloperDiskImage.dmg for iOS 17 not found in Xcode 15

### ISSUE THREAD

**Issue by vashworth** [ios-control/ios-deploy#588]
As of Xcode 15 and iOS 17, DeveloperDiskImage.dmg is no longer found in the expect path, which is Xcode.app > Contents > Developer > Platforms > iPhoneOS.platform > DeviceSupport > 17.x.

This appears to be a permanent change:
https://developer.apple.com/forums/thread/730947
https://github.com/libimobiledevice/libimobiledevice/issues/1452

# Expected behavior
Debugging starts properly

# Actual behavior.
Debugging phase fails due to missing DeveloperDiskImage.dmg.
```
------ Debug phase ------
Starting debug of xxxx (J617AP, J617AP, uknownos, unkarch, 17.0, 21A5248v) a.k.a. 'xxx' connected through USB...
2023-06-09 11:42:16.219 ios-deploy[45135:1129696] [ !! ] Unable to locate DeviceSupport directory with suffix 'DeveloperDiskImage.dmg'. This probably means you don't have Xcode installed, you will need to launch the app manually and logging output will not be shown!
```

# Steps to reproduce the problem
1. Download Xcode 15 beta
2. Select new Xcode 
```
xcrun xcode-select --switch /Applications/Xcode-beta.app/Contents/Developer
```
3. Download iOS 17 on a physical device
4. Run ios-deploy
```
ios-deploy --id xxxx --bundle path/to/Runner.app --app_deltas path/to/app-delta --debug --no-wifi 
```

# System Specs

Please run the commands below in your Terminal.app and include it in the issue. Check when done and include results below.

- [x] 1. system_profiler SPSoftwareDataType
```
Software:

    System Software Overview:

      System Version: macOS 13.4 (22F66)
      Kernel Version: Darwin 22.5.0
      Boot Volume: Macintosh HD
      Boot Mode: Normal
      Computer Name: xxx
      User Name: xxx
      Secure Virtual Memory: Enabled
      System Integrity Protection: Enabled
      Time since boot: 1 day, 6 hours, 59 minutes
```
- [x] 2. ios-deploy -V
`1.12.0`
- [x] 3. xcodebuild -version
```
Xcode 15.0
Build version 15A5160n
```
- [x] 4. xcode-select --print-path
`/Applications/Xcode-beta.app/Contents/Developer`
- [x] 5. gcc --version
```
Apple clang version 15.0.0 (clang-1500.0.28.1.1)
Target: arm64-apple-darwin22.5.0
Thread model: posix
InstalledDir: /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```
- [x] 6. lldb --version
```
lldb-1500.0.14.6
Apple Swift version 5.9 (swiftlang-5.9.0.114.6 clang-1500.0.27.1)
```

**Comment by gabebear** [comment]
I'm not sure if we want to have ios-deploy work with Xcode15; it's possible Xcode14 will be the final version supported. Apple has FINALLY created a non-hack mechanism with `devicectl`(e.g. try running `xcrun devicectl --help`). https://developer.apple.com/documentation/Updates/Xcode

**Comment by steveames** [comment]
Even using xcrun to install/launch there's still the issue of connecting to lldb and managing stderr/stdout. That's the majority of what I use ios-deploy for. I think that purpose is still legit in xcode15-land 
unless there's a simple command for that as well that should go in the readme?

**Comment by gabebear** [comment]
It does seem there is still a need for a way to launch and connect to lldb. Ideally Apple will add this to devicectl.

Apple tracks feedback/radar reports to gauge interest in products. People who need iOS device debugging scripting support should file requests via https://feedbackassistant.apple.com/ . This could possibly get Apple to add this before the team that created devicectl moves to other things.

**Comment by Saafo** [comment]
> ```bash
> xcrun devicectl device install app --device 00008110-001XXXXXXXXXX ./xgen/Build/Products/Release-iphoneos/nilo.app
> crun devicectl device process launch --device 00008110-001XXXXXXXXXX file:///private/var/containers/Bundle/Application/1604D2D5-35F3-4E43-8B47-1DEF5D778480/nilo.app
> ```
Seems Apple's devicectl is not easy to use, can we wrap it under the hood after Xcode 15?

**Comment by steveames** [comment]
> Ideally Apple will add a way to connect to lldb to devicectl.
Yep. That'd be pretty nice. With CoreDevice there are IPv6 tunnels created to each device. You can point lldb directly at one of those and connect to it. The filehandle/socket forwarder is no longer required. The trick is figuring out the port that the debugserver is running on. That seems a tad convoluted currently. Doable but convoluted. I did take your advice and submit a feedback suggestion.

**Comment by SachaTe** [comment]
>
```
xcrun devicectl device install app --device 00008110-001XXXXXXXXXX ./xgen/Build/Products/Release-iphoneos/nilo.app
xcrun devicectl device process launch --device 00008110-001XXXXXXXXXX file:///private/var/containers/Bundle/Application/1604D2D5-35F3-4E43-8B47-1DEF5D778480/nilo.app
```
`devicectl` is not as simple to use, we print some results to std::out and it seems like there is no way to capture this data like ios-deploy used to do. Also, we are stuck installing, launching, and then continuously polling the device to see if the application ran, as there is no way to attach to process launch

In addition with `ios-deploy` I was able to easily upload files to the bundle and it seems like with `devicectl` I keep getting errors

**Comment by sk2212** [comment]
> > 
> 
> ```
> xcrun devicectl device install app --device 00008110-001XXXXXXXXXX ./xgen/Build/Products/Release-iphoneos/nilo.app
> xcrun devicectl device process launch --device 00008110-001XXXXXXXXXX file:///private/var/containers/Bundle/Application/1604D2D5-35F3-4E43-8B47-1DEF5D778480/nilo.app
> ```
> 
> `devicectl` is not as simple to use, we print some results to std::out and it seems like there is no way to capture this data like ios-deploy used to do. Also, we are stuck installing, launching, and then continuously polling the device to see if the application ran, as there is no way to attach to process launch
> 
> In addition with `ios-deploy` I was able to easily upload files to the bundle and it seems like with `devicectl` I keep getting errors

With the latest pymobiledevice3 version from https://github.com/doronz88/pymobiledevice3 I have found a method how to establish remote debugging with an iPhone with iOS 17:

**Create a rsd-Tunnel:**
`sudo python3 -m pymobiledevice3 remote start-tunnel --script-mode
`
-> will output the rsd data ($rsd_output)

**Start debugserver on iPhone**
`python3 -m pymobiledevice3 developer debugserver start-server --rsd $rsd_output`

(e.g. `python3 -m pymobiledevice3 developer debugserver start-server --rsd fdf0:721a:e29b::1 57537`)

This will print something like this:

```
(lldb) platform select remote-ios
(lldb) target create /path/to/local/application.app
(lldb) script lldb.target.module[0].SetPlatformFileSpec(lldb.SBFileSpec('/private/var/containers/Bundle/Application/<APP-UUID>/application.app'))
(lldb) process connect connect://[fdf0:721a:e29b::1]:54437   <-- ACTUAL CONNECTION DETAILS!
(lldb) process launch

```
Using `xcrun devicectl device install app` command to install the App from an archive will let us parse out the data which is needed to setup lldb:


```
install@DE011-CMXXXX Scripts % xcrun devicectl device install app --device 00008110-xxxxxxx /Users/install/Projects/build_http_communication_Qt/Release-iphoneos/QtTest.app
12:04:26  Acquired tunnel connection to device.
12:04:26  Enabling developer disk image services.
12:04:26  Acquired usage assertion.
42%... 57%... 60%... 62%... 66%... 68%... 72%... 74%... 76%... 80%... 84%... 88%... 92%... 96%... Complete!
App installed:
? bundleID: de.bender.qtcomponenttest
? installationURL: file:///private/var/containers/Bundle/Application/8D43DD90-B138-4C42-8BE8-2752CAB6A259/QtTest.app/
? launchServicesIdentifier: unknown
? databaseUUID: 5D6368B1-5A96-462B-B5AD-B9F12163FEEA
? databaseSequenceNumber: 1508
? options:
```

With this output we can setup lldb:

```
$ lldb
(lldb) platform select remote-ios
(lldb) target create /Users/install/Projects/build_http_communication_Qt/Release-iphoneos/QtTest.app
(lldb) script lldb.target.module[0].SetPlatformFileSpec(lldb.SBFileSpec('/private/var/containers/Bundle/Application/8D43DD90-B138-4C42-8BE8-2752CAB6A259/QtTest.app'))
(lldb) process connect connect://[fdf0:721a:e29b::1]:54437
(lldb) process launch

```

Will result in an starting iOS app with debug output to lldb.

Next step is to pipe this output to a file, std_out or something else.

Maybe someone how knows ios-deploy is able to integrate this stuff. 

I end up in creating a zsh script which is creating and parsing all the stuff to get the steps above automated.

**Comment by megabitsenmzq** [comment]
> I end up in creating a zsh script which is creating and parsing all the stuff to get the steps above automated.

Could you share your script? It will be so helpful.

**Comment by laiyierjiangsu** [comment]
@sk2212 

hi, sk2212, 

I have some questions about your script.

1. 
 I don't understand the code "curl -sS "127.0.0.1:5555"" for . On my mac, failed to connect . where is 5555 coming from ?

`
++ echo
+ device_connected=
+ '[' '' '!=' '' ']'
+ curl -sS 127.0.0.1:5555
curl: (7) Failed to connect to 127.0.0.1 port 5555 after 0 ms: Couldn't connect to server
+ exit_code=
`
2.  echo ${sudo_pw} | sudo -S python3 -m pymobiledevice3 remote tunneld > server_start.o 
On my mac , there is nothing output to server_start.o .  On the other hand , I found server_start.o  have no further using in the whole script .how do you get  the rst output ? 

I am the first time to use mac real machine debugging, please forgive my little white.

**Comment by laiyierjiangsu** [comment]
Can anyone knows "127.0.0.1:5555" in sk2212's script? I really don't understand it's purpose. If someone knows,please share with me . I really appreciate.

**Comment by sk2212** [comment]
> Can anyone knows "127.0.0.1:5555" in sk2212's script? I really don't understand it's purpose. If someone knows,please share with me . I really appreciate.

curl -sS "127.0.0.1:5555" is just to check if there is a response provided by tunnel server instance started with pymobiledevice3. Do you have pymobiledevice3 installed?

**Comment by laiyierjiangsu** [comment]
> > Can anyone knows "127.0.0.1:5555" in sk2212's script? I really don't understand it's purpose. If someone knows,please share with me . I really appreciate.
> 
> curl -sS "127.0.0.1:5555" is just to check if there is a response provided by tunnel server instance started with pymobiledevice3. Do you have pymobiledevice3 installed?
@sk2212 
Yes. I have installed pymobiledevice3, and following your instructions step by step, I am able to launch and debug apps installed on the phone using 'process launch'. However, when I try to use the script you provided to simplify the whole process, I find that 'server_start.o' is empty, resulting in an empty 'rsd_output'.

Given the presence of port 5555, I suspect this is a service launched locally by pymobiledevice3, but I couldn't find any documentation about this port in the pymobiledevice3 repository. I would be very grateful if you could provide me with the documentation for port 5555. Thank you very much.

**Comment by sk2212** [comment]
@laiyierjiangsu yeah, it seems that the standard port '5555' was changed in 'pymobiledevice3':

https://github.com/doronz88/pymobiledevice3/pull/874/commits/6b0145eea685fd02a88ca51b0fe68bfc16e6706a

**Comment by laiyierjiangsu** [comment]
@sk2212 Thank you for your help. I think I should read through the source code of pymobiledevice3 to thoroughly understand its mechanisms. It looks very interesting."

**Comment by jswillard** [comment]
For anyone still wanting to deploy and debug from the command line, Apple added support natively to LLDB in Xcode 16 Beta. See my comment on the flutter repo here: https://github.com/flutter/flutter/issues/133465#issuecomment-2159512125

**Comment by gitevedev** [comment]
Considering [these significant drawbacks](https://github.com/ios-control/ios-deploy/issues/588#issuecomment-1814679642) of the new `devicectl`, comparing to `ios-deploy`, wondering how hard would it be to update `ios-deploy` to work with the newer iOS? I understand the whole device communication framework was changed by Apple, and it's not going to be an easy task?

**Comment by rolfbjarne** [comment]
> Considering [these significant drawbacks](https://github.com/ios-control/ios-deploy/issues/588#issuecomment-1814679642) of the new `devicectl`,

`devicectl` can capture stdout/stderr and wait for the app to exit now:

```shell
$ xcrun devicectl help device process launch
[...]
--console               Attaches the application to the console and waits for it to exit.
        devicectl will wait for the app to terminate. Catchable signals sent to devicectl are forwarded to the app. If the app is not already running, its standard streams will be connected to devicectl's standard streams.
```

Regarding the other problem (errors uploading files to the device), it seems it would be easier to report any issues to Apple, instead of re-implementing the entire tool?
